### ISSUE TITLE
ios/safari audioContext with wrong sampleRate 

### ISSUE THREAD

**Issue by electromotif** [goldfire/howler.js#1141]
We're using assets in 44.1 kHz, playback in 48kHz results in crackles/pops in audio. 

Observed on iphone X running ios 12.1.4, Howler 2.1.1

The unload and creation of a new audioContext does not seem to cure it this time.
Safari on an ipad air2 running same ios release creates an audioContext with a 44.1kHz rate.
Also interesting is that playback on headphones does not have the crackling.

related to issue 434
Also see
https://bugs.webkit.org/show_bug.cgi?id=154538
https://bugs.webkit.org/show_bug.cgi?id=168165

**Comment by st-h** [comment]
Commenting because I think I am currently searching for a similar solution and the symptoms you describe seem very related - however, I stopped using howler very recently because the needs of our app have changed quite a bit.

My issue is described in detail here https://stackoverflow.com/questions/54982847/crackles-and-noises-when-playing-ios-web-audio-with-simple-graph 

I also opened a bug within the apple bug tracker today

What I observed is:
- playback issues appear on iOS devices only (works fine on android and macOS)
- issues are independent of the browser being used (tried safari and chrome)
- issues are independent of the sample rate being used (iPad Pro consistently reports 44.1kHz, while iPhone 6s reports 48kHz - the issue can be reproduced on both at both sample rates with the same audio file)

Are you observing the degraded audio playback all the time, or does it sometimes work correctly?

**Comment by miltoncandelero** [comment]
Just to add more inconclusive info:

We are working on a game using the OpenFL toolchain, but audio is achieved using howler so it applies.

We got feedback that "after a while" the game sounds crackle and get all _statick-y_ on an iPhone 7 (11.4) and iPhone 7 (12.1). We couldn't recreate the issue on our old AF ipad.

We found that we had 2 soundfiles that had a sample rate really low and didn't match the rest of the files (all other files where 44.1khz)

We normalized the sample rate (to 44.1) and for good measure the bitrate too (128kbps) and are waiting for more feedback. Will update as we get more info.

**Comment by electromotif** [comment]
> Are you observing the degraded audio playback all the time, or does it sometimes work correctly?

We get it on our testing/qa Iphone X, and like you noticed, on Safari and Chrome. 

On another Iphone X, running the same ios version, it does not appear. 

iphone 7, XR and ipad is fine. At this time. :o)

**Comment by miltoncandelero** [comment]
Ok, anyone has any idea what is going on? I am panicking a little here.

I had the QA team test the sample tests from https://howlerjs.com/assets/howler.js/tests/ and they tell me that EVERYTHING sounds glitchy. Webaudio and html5 audio alike. (on an iPhone X)

**Comment by IriySoft** [comment]
Same issue here. Any ideas, anyone?..

**Comment by miltoncandelero** [comment]
Still having issues with this. Please, did anybody find any workarround?

**Comment by jacobclarke92** [comment]
Big oof on this one being a dead end so far, but I'll add my experiences.
My team did work out it was caused by sample rate mismatch between 48k and 44.1k -- this mismatch can be across audio file played in or out of howler, and also includes videos.

If anyone needs to do this in bulk we wrote a quick script to make this sanity checking and fixing more manageable because we have heaps of audio files (requires mac and ffmpeg)
```bash
#!/bin/bash


linebreak="\
-----------------------------"

# get folder path
echo -n "Drag a folder to recursively scane for and fix MP3s not set to 44100 sample rate and press [ENTER]: "
read folderpath
echo "$folderpath"

# check if entered path is a folder
if [ -d "${folderpath}" ]; then
    
    # set foldername and basefolder vars
    foldername=`basename "${folderpath}"`
    cd "${folderpath}"
    
    find . -name "*.mp3" | while read fname; do
        if [[ ! $fname == *"_old.mp3"* ]]; then
            sampleRate=`ffprobe -v error -show_format -show_streams "${fname}" | grep sample_rate`
            if [ "$sampleRate" != "sample_rate=44100" ]; then
                echo "$fname"
                echo "$sampleRate"
                newName="${fname}_old.mp3"
                exec `mv "${fname}" "${newName}"`
                echo "ffmpeg -i \"${newName}\" -ar 44100 \"${fname}\""
                exec `ffmpeg -i "${newName}" -ar 44100 "${fname}" > /dev/null 2>&1`
                echo "${linebreak}"
            fi
        fi
    done
    
fi
```

Now that fixed the original "crackling" issue, however we're now experiencing a new issue that sounds like bitcrushing or distortion exclusively on iPad Pro so far ... fun times.

Chime in if any of you have any new insights after all this time!

**Comment by jespertheend** [comment]
I'm not using howler but I am using AudioContext and seeing a similar issue. All our files are 48khz.
I still haven't been able to fix this. It seems to vary between builds so I haven't been able to pin point the exact issue. One thing I do know is that I can consistently reproduce the issue by opening the web page while headphones are connected, and then disconnecting the headphones.

**Comment by Ronsku** [comment]
I'm getting crackles from following setup / device / browser combinations and there seems to be no way around it:
All combinations of:
**Device:** iPhone 5, (iOS 10), iPhone 6 (iOS 12), iPhone 11 Pro (iOS 13)
**Sample rate:** 22.05khz, 44.1khz, 48khz, 88.2khz
**Bit Rate:** 96kbit/s, 128kbit/s, 320kbit/s
**File formats:** mp3, ogg
**Audio file:** Separate & Sprited
**Howler.js Version:** 2.1.3

Every time I get crackling sounds, no matter how I configure my sound file. Is there some configuration I can do on Howler's side to match it up with the sound files?

This is a huge problem with the sound engine, is someone looking into this?

**Comment by themoonrat** [comment]
I don't know if this is the exact issue you guys are having, but there is a known issue with iOS not resetting it's sample rate correctly

https://github.com/Jam3/ios-safe-audio-context

**Comment by miltoncandelero** [comment]
> I don't know if this is the exact issue you guys are having, but there is a known issue with iOS not resetting it's sample rate correctly, which produces the same issues you are seeing.
> 
> https://github.com/Jam3/ios-safe-audio-context
> 
> I use a variant of that hack each time the content becomes 'active' again

@goldfire Any chance this can be implemented into howler? or maybe a similar workaround?
(Sorry I can't pr myself, my js is really crappy :( )

**Comment by paulmasri** [comment]
Just in case it's relevant, [over on the Steinberg forum they're talking about iOS sample rate issues and crackling](https://www.steinberg.net/forums/viewtopic.php?t=153295#p823404), which they put down to the increased CPU required for sample rate conversion.

**Comment by rvanderspuy** [comment]
Having the same problem!

**Comment by jpjokinen** [comment]
https://www.steinberg.net/forums/viewtopic.php?t=153295#p823404

"The reason for this change is that the automatic sample rate conversion by iOS results in increased CPU usage, which would have lead to issues like crackling on the iPad Pro 3rd gen. models.

The available sample rates are dictated by the hardware of the iPad device in use. For example, the 2018 iPad Pro models (3rd gen) only support 48kHz with the internal speakers and mic, unless external hardware is connected, whereas other iPads support 44.1 and 48 kHz."

**Comment by dbFlower** [comment]
any workaround yet? having the same problem.

**Comment by ErikSom** [comment]
I also would like to find a proper work around for this one

**Comment by fknorn** [comment]
Yep, same here :( Any workaround found? I'm getting this crackling sound when using NoSleep.js together with Howler...

**Comment by dbFlower** [comment]
已收到
